src/fenwick.jl

```diff
@@ -0,0 +1,96 @@
+"""
+    FenwickTree{T}(n)
+    
+Constructs a [`FenwickTree`](https://en.wikipedia.org/wiki/Fenwick_tree) of length `n`.
+ 
+"""
+mutable struct FenwickTree{T}
+    bit::Vector{T}
+    n::Integer
+end
+
+FenwickTree{T}() where T = FenwickTree{T}(0)
+FenwickTree{T}(n::Integer) where T = FenwickTree{T}(zeros(T, n), n)
+
+"""
+    FenwickTree(arr::AbstractArray) 
+    
+Constructs a [`FenwickTree`](https://en.wikipedia.org/wiki/Fenwick_tree) from `arr`.
+ 
+"""
+function FenwickTree(a::AbstractVector{U}) where U
+    n = length(a)
+    bit = FenwickTree{U}(n)
+    @inbounds for i = 1:n
+        update!(bit, i, a[i])
+    end
+    bit
+end
+
+bit(F::FenwickTree{T}) where T = F.bit
+length(F::FenwickTree{T}) where T = F.n
+
+"""
+    update!(F::FenwickTree{T}, ind, val::T)
+
+Update the value of the [`FenwickTree`] by `val` from the index `ind` upto the length of the Fenwick Tree.
+
+""" 
+function update!(F::FenwickTree{T}, ind::Int, val::T) where T
+    i = ind
+    n = F.n
+    @boundscheck 1 <= i <= n || throw(ArgumentError("$i should be in between 1 and $n"))
+    @inbounds while i <= n
+        F.bit[i] += val
+        i += i&(-i)
+    end
+end
+
+"""
+    update!(F::FenwickTree{T}, left, right, val::T)
+
+Update the value of the [`FenwickTree`] by `val` from the index `left` upto the index `right`.
+
+"""    
+function update!(F::FenwickTree{T}, left::Int, right::Int, val::T) where T
```

Comment:
If updated to `inc!`, consider changing this to 

```julia
function inc!(ft::FenwickTree{T}, range::AbstractUnitRange, val::T) where T
```

(with appropriate changes to the function below.)

You also might consider dropping the `T` and instead converting val to type `T` below, to give a user more flexibility.

See, e.g., [here](https://github.com/JuliaLang/julia/blob/master/base/dict.jl#L371-L372) for an example of this usage.